src/app/vm/vm-sidebar/affinity-group-selector/affinity-group-selector.component.ts

```diff
@@ -1,80 +1,101 @@
+  @Input()
   public affinityGroups: AffinityGroup[];
+  @Input()
+  public sortedAffinityGroups: AffinityGroup[];
+  @Input()
+  public preselectedAffinityGroups: AffinityGroup[];
+  @Input()
+  public enablePreselected: boolean;
+  @Output()
+  public createdAffinityGroup = new EventEmitter<AffinityGroup>();
+  @Output()
+  public submited = new EventEmitter<string[]>();
+  @Output()
+  public canceled = new EventEmitter();
   public loading: boolean;
-  public selectedAffinityGroupName: string;
-  public vm: VirtualMachine;
-
-  constructor(
-    private affinityGroupService: AffinityGroupService,
-    private dialogRef: MatDialogRef<AffinityGroupSelectorComponent>,
-    private dialogService: DialogService,
-    @Inject(MAT_DIALOG_DATA) public data,
-  ) {
-    this.vm = data.vm;
-  }
+  public selectedGroup: AffinityGroup;
+  public types = [AffinityGroupType.antiAffinity, AffinityGroupType.affinity];
+  public affinityGroupForm: FormGroup;
+  public maxEntityNameLength = 63;
 
-  public ngOnInit(): void {
-    this.loadGroups();
-  }
+  constructor(private formBuilder: FormBuilder) {}
 
-  public get groupName(): string {
-    return this.vm.affinitygroup.length && this.vm.affinitygroup[0].name;
+  public ngOnChanges(changes: SimpleChanges) {
+    const affinityGroup = changes.affinityGroups.currentValue;
+    if (affinityGroup) {
+      this.affinityGroups = affinityGroup;
+      this.createForm();
+    }
   }
 
-  public get groupNames(): string[] {
-    return this.affinityGroups.map(_ => _.name);
-  }
-
-  public get selectedAffinityGroup(): AffinityGroup {
-    return this.affinityGroups.find(_ => _.name === this.selectedAffinityGroupName);
-  }
-
-  public createGroup(name: string): void {
-    this.affinityGroupService
-      .create({
-        name,
-        type: AffinityGroupType.hostAntiAffinity,
-      })
-      .subscribe(
-        affinityGroup => this.dialogRef.close(affinityGroup.id),
-        error => this.dialogService.alert({ message: error.message }),
-      );
+  public ngOnInit(): void {
+    this.createForm();
   }
 
-  public changeGroup(name: string): void {
-    this.selectedAffinityGroupName = name;
-    this.dialogRef.close(this.selectedAffinityGroup.id);
+  public createGroup(): void {
+    const newAffinityGroup = {
+      name: this.affinityGroupForm.value.name,
+      type: this.affinityGroupForm.value.type,
+      description: this.affinityGroupForm.value.description,
+    } as AffinityGroup;
+    this.affinityGroupForm.reset();
+    this.createdAffinityGroup.emit(newAffinityGroup);
   }
 
-  public removeGroup(): void {
-    this.dialogRef.close('');
+  public changeGroup(group: AffinityGroup): void {
+    this.selectedGroup = group;
   }
 
-  public onCancel(): void {
-    this.dialogRef.close();
+  public submit(): void {
+    let selectedGroupIds;
+    if (this.enablePreselected) {
+      selectedGroupIds = [this.selectedGroup.id];
```

Comment:
You can simplify this function by extracting `getSelectedAffinityGroups`, so you don't have to use `let`.

```
if (this.enablePreselected) {
  return [this.selectedGroup.id];
else if (this.selectedGroup.id === emptyAffinityGroup) {
  return [];
else
...
```